### PR TITLE
COMMERCE-5270 Simplified QuantitySelector with a bunch of fixes

### PR DESCRIPTION
**Pull request by gianmarcobrunialti**
This PR fixes a bunch of things that are all correlated with the latest updates of the QuantitySelector components, which include:
* Broken component styles in MiniCart (COMMERCE-5270)
* Non-updating quantity value on interaction (COMMERCE-5271)
* Component loop re-renders
* Performance killing `select > option` element rendering when there are multiple instances of the component (worse case scenario being the Catalog page *with* subsequent MiniCart load-up/mount/unmount which may potentially contain loads of `QuantitySelector`'s)

What's inside:
* In the MiniCart we're always going to use a `ClayInput` number-typed QuantitySelector (as agreed with Andrea Censi), which is correctly debounced to avoid keypress overhead upon quantity update (and subsequent API call). Being a `number`-typed input it inherits the full keyboard accessibility and `multipleQuantity` settings 
* In the ProductCard (soon to be) Andrea would like to use the ClaySelect, which can be enabled now via prop or quantity settings.
* In the MiniCart the ClayInput will be switched to a `ClaySelect` if appropriate quantity settings are passed in (e.g. `allowedQuantities`)

### CODE REVIEW

**Comment by gianmarcobrunialti** [comment]
ci:test:sf

**Comment by gianmarcobrunialti** [comment]
ci:test:relevant

**Comment by gianmarcobrunialti** [comment]
ci:test:sf

**Comment by wincent** [comment]
ci:test:commerce

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js [hunk 1]
```diff
@@ -12,175 +12,115 @@
...
-							<ClaySelect.Option
-								key={item}
-								label={item}
-								value={item}
-							/>
-						))}
-					</ClaySelect>
-				)}
-			</div>
+		TIMEOUT_REF = setTimeout(
```
Comment (wincent) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js:62]: This shouldn't be all-caps; we usually use all-caps for _constants_.

Also, without really understanding this part of the code base, can there ever be more than one `QuantitySelector` on screen at a time? Seeing as they would be sharing the same `TIMEOUT_REF` reference... is that what you want?

Might be more "hygienic" to stick this in a `useRef` reference which is tied to the component instance.

```js
const timeout = useRef();

// ... and later...

timeout.current = setTimeout(...);
```

File: modules/apps/commerce/commerce-frontend-js/test/dev/components/QuantitySelector.js [hunk 2]
```diff
@@ -17,12 +17,14 @@ import launcher from '../../../src/main/resources/META-INF/resources/components/
...
-	settings: {
-		maxQuantity: 29,
-		minQuantity: 1,
-	},
-	size: 'large',
+	onUpdate: (quantity) =>
+		new Promise((resolve) => {
+			setTimeout(() => {
+				// eslint-disable-next-line no-console
+				console.log('The new quantity is:', quantity);
```
Comment (wincent) [inline modules/apps/commerce/commerce-frontend-js/test/dev/components/QuantitySelector.js:24]: Please don't log this (see [the guidelines on logging](https://github.com/liferay/liferay-frontend-guidelines/blob/master/dxp/debugging.md#logging)). If you _must_ log it, guard it behind an `if (process.env.NODE_ENV === 'development')` check.

File: modules/apps/commerce/commerce-frontend-js/test/quantitySelector/utils/index.js [hunk 3]
```diff
@@ -14,47 +14,73 @@
...
 import * as Utils from '../../../src/main/resources/META-INF/resources/components/quantity_selector/utils/index';
 
-describe('generateOption`s working', () => {
-	it('display allowedQuantities as options if this is not empty', () => {
-		const allowed = [2, 4, 65, 33, 913, 267, 323, 122, 90, 113];
-		expect(Utils.generateOptions({allowedQuantity: allowed})).toEqual(
-			allowed
+describe('QuantitySelector Util -> generateQuantityOptions', () => {
+	it('returns allowedQuantities, if non-empty, as formatted options', () => {
+		const allowedQuantities = [2, 4, 65, 33, 913, 267, 323, 122, 90, 113];
```
Comment (wincent) [inline modules/apps/commerce/commerce-frontend-js/test/quantitySelector/utils/index.js:19]: What about 42?

just-messing

**Review by wincent — COMMENTED** [review]
lg2m apart from a couple of small things noted inline. The only "must-fix" in there would be the `console.log`; the rest I'd leave up to you @gianmarcobrunialti. Nice tests too!

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js [hunk 4]
```diff
@@ -12,175 +12,115 @@
...
+	forceDropdown,
+	name,
+	onUpdate,
+	quantity,
+	...optionSettings
+}) {
+	const [selectedQuantity, setSelectedQuantity] = useState(
+		optionSettings.minQuantity > quantity
+			? optionSettings.minQuantity
+			: quantity
```
[the remark this answers is not in the document]
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js:35]: Definitely cleaner, didn't change it from previous iteration. I'll change it!

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js (hunk 1 above)
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js:62]: Agreed for the useless caps. Yes, there will be multiple instances of quantity selectors in the page, and I didn't take the shared ref as an issue because, at the end of the day, it won't be possible to change > 1 Input/Select value at the same time. So by the time someone interacts with one of these component, I was just preferring to reuse the same "garbage" var not to pollute the component itself. What do you think? @wincent

File: modules/apps/commerce/commerce-frontend-js/test/dev/components/QuantitySelector.js (hunk 2 above)
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/test/dev/components/QuantitySelector.js:24]: This component "starter" won't ever run in Liferay, it lives in our static environment where components are rendered out of context in a blank page. Is this a margin that's reasonable enough to leave this `console.log` where it is? 😄  If not, I'll figure out something else 👍

File: modules/apps/commerce/commerce-frontend-js/test/quantitySelector/utils/index.js (hunk 3 above)
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/test/quantitySelector/utils/index.js:19]: I took that array from a previous iteration. Incredibly, there are living beings in this world who don't take towels to be the _most massively useful thing_, let alone understanding the ultimate über value of answers like "42". I'mma fix this havoc.

**Comment by gianmarcobrunialti** [comment]
> lg2m apart from a couple of small things noted inline. The only "must-fix" in there would be the `console.log`; the rest I'd leave up to you @gianmarcobrunialti. Nice tests too!

🙇

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js (hunk 1 above)
Comment (wincent) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js:62]: The reason "components" exist as a concept at all is for resuability. So, _even if they aren't actually reused in practice_, designing them for reusability makes them better.

As such, I would definitely make the component re-use-friendly by avoiding invisible shared global state. And while you say it's not possible to change more than one value at a time right now, making this change will make your code more resilient against possible changes in the future, if the constraints ever change (and they often do...)

tl;dr: It's not pollution, it's encapsulation.

File: modules/apps/commerce/commerce-frontend-js/test/dev/components/QuantitySelector.js (hunk 2 above)
Comment (wincent) [reply modules/apps/commerce/commerce-frontend-js/test/dev/components/QuantitySelector.js:24]: Ah, I didn't even notice the path. No need to worry about the `NODE_ENV` check then, but I think you should just get rid of the `console.log` because lint suppressions _are_ a smell. The more we normalize them, the more people get used to seeing them and before you know it the repo is littered with `console.log` and suppressions and we have a mess.

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js (hunk 1 above)
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js:62]: Oh btw, it's a short keypress debounce, if you're wondering.

Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js:62]: @wincent Alright! 👍

File: modules/apps/commerce/commerce-frontend-js/test/dev/components/QuantitySelector.js (hunk 2 above)
Comment (gianmarcobrunialti) [reply modules/apps/commerce/commerce-frontend-js/test/dev/components/QuantitySelector.js:24]: Cool, cool.

**Comment by gianmarcobrunialti** [comment]
ci:test:sf

**Comment by gianmarcobrunialti** [comment]
ci:test:relevant

**Comment by ivicac** [comment]
ci:test:relevant

**Comment by ivicac** [comment]
ci:test:relevant

**Comment by gianmarcobrunialti** [comment]
ci:test:relevant

**Comment by gianmarcobrunialti** [comment]
Jenkins gave a `BUILD SUCCESSFUL` for `relevant` tests, re-triggering to label it via automation.

**Review by marco-leo — APPROVED** [review]

**Comment by ivicac** [comment]
ci:forward

**Comment by gianmarcobrunialti** [comment]
ci:forward

**Comment by ivicac** [comment]
ci:test:sf

**Comment by ivicac** [comment]
ci:test:relevant

**Comment by ivicac** [comment]
ci:forward
